### ISSUE TITLE
path in url is not encoded, list fails when path contains a "%"

### ISSUE THREAD

**Issue by sls-cat** [hvac/hvac#248]
I0815 16:21:24.884774 140736674636736 connectionpool.py:393] https://vault-dev.xxx.net:8200 "GET /v1/foo/bar/client-secret/%1?list=True HTTP/1.1" 400 None

The hvac list function (and probably other functions) fails with python 3.6.5 and hvac 0.6.1 and 0.6.3.  The vault 0.10.2 command line displays output correctly.

      File "/usr/local/lib/python3.6/site-packages/hvac/v1/__init__.py", line 54, in list
        return self._get('/v1/{}'.format(path), params=payload).json()
      File "/usr/local/lib/python3.6/site-packages/hvac/v1/__init__.py", line 1589, in _get
        return self.__request('get', url, **kwargs)
      File "/usr/local/lib/python3.6/site-packages/hvac/v1/__init__.py", line 1640, in __request
        self.__raise_error(response.status_code, text, errors=errors)
      File "/usr/local/lib/python3.6/site-packages/hvac/v1/__init__.py", line 1646, in __raise_error
        raise exceptions.InvalidRequest(message, errors=errors)
    hvac.exceptions.InvalidRequest: 400 Bad Request

    sls@mac 16:25> vault list /foo/bar/client-secret
    Keys
    ----
    %1/
    live/
    sls@mac 16:25> vault list /foo/bar/client-secret/'%1'
    Keys
    ----
    %2/
    sls@mac 16:25>

**Comment by jeffwecan** [comment]
Thanks for the report @sls-cat! I'll aim to make sure we get this corrected in the next release (0.6.4).

**Comment by jeffwecan** [comment]
For what it is worth, I'm working through a fix for this issue now. (Fair warning, it would be helpful to have you validate the proposed fix once it gets sorted if you're willing @sls-cat, so I may be tagging you on an associated PR in short order.)

**Comment by sls-cat** [comment]
I'm happy to take a look at the diff and test it.  It would be good if
requests lib could handle the url quoting.  I like to leave the HTTP specs
to people who write those libraries.

**Comment by jeffwecan** [comment]
@sls-cat: I'm having a bit of trouble nailing down a regression test case for this issue. This is what I sketched out thus far:

```python
    def test_list_path_with_url_encoded_string(self):
        test_data_value = 'regression_test_gh_issue_248'
        self.client.write('secret/%1/test-path', test=test_data_value)
        list_secret_response = self.client.list('secret/%1')
        self.assertIn(
            member='test-path',
            container=list_secret_response['data']['keys'],
        )
```

As-is, the the above assertion passes. From the example request log in the initial issue description ("GET /v1/foo/bar/client-secret/%1?list=True HTTP/1.1"), I would have expected to hit the same `hvac.exceptions.InvalidRequest` exception. For what it is worth, the `list_secret_response` variable in the above example test case is ending up with these contents:

```python
pprint(list_secret_response)
{u'auth': None,
 u'data': {u'keys': [u'test-path']},
 u'lease_duration': 0,
 u'lease_id': u'',
 u'renewable': False,
 u'request_id': u'5af99f1a-05e4-a659-de4e-91c5cbd27f4f',
 u'warnings': None,
 u'wrap_info': None}
```

I'll try to look into this a bit more later to see if I'm misunderstanding some aspect. However if you are able to provide explicit replication steps (e.g., 1.) write a path of "some-path" to Vault 0.10.2 using vault CLI 2.) Instantiate an hvac Client and call the `list()` method with the following args [...], etc. that may be a great help.

**Comment by briantist** [comment]
Seems to be fixed by #532
